### PR TITLE
LPS-166479 - Create new client extension for theme spritemap

### PR DESCRIPTION
**Pull request by bryceosterhaus**
This PR adds a new field to themeDisplay for accessing the spritemap and also adds a client extension to override this value. Icons are tied pretty closely to the theme, as is JS and CSS and so it didn't make much sense to continue with the "icon pack" idea as our client extensions are overriding these sorts of theme elements. I wasn't able to come up with a great solution that would both allow for icon packs and client extension for icon pack, they seemed to address the same concern in different ways. By moving the spritemap to themeDisplay we now have a singular entry point to access icons and also aligns the resource similarly to our css and js theme resources.

## How to test:
- Add FF "LPS-166479"
- Upload new SVG to document library
- Copy URL to file
- Create "SVG" client extension using the URL from the doc-lib
- Navigate to edit the "Look and Feel" of you page
- Select the registered "Theme Spritemap Client Extension"
- Save
- Go back to page and see new SVG used as spritemap
	- You can even inspect the page and see the the icons reference the spritemap in the document library

Here is an SVG file to use, I simply took the existing spritemap and swapped "cog" for "pencil". [icons.svg](https://user-images.githubusercontent.com/6843530/204792992-bed2c337-d017-40f5-ad13-5917ee5b129e.svg)

### Video Demo
https://user-images.githubusercontent.com/6843530/204792878-8ebc8921-6ec8-4aa3-b283-a0060a85920b.mp4

### CODE REVIEW

**Comment by bryceosterhaus** [comment]
cc @dsanz @izaera if you have any feedback, let me know

**Comment by mateusralv** [comment]
I'm taking a look 👓

**Comment by mateusralv** [comment]
Hi, @bryceosterhaus I found some unique failures in your PR, when you have time, could take a look? Thanks 😄 

[comparison run ](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1669835634333&testrayRunIdA=2089032666&testrayRunIdB=2089627718)
### Unique failures
1. [modules-semantic-versioning-jdk8/0/0](https://testray.liferay.com/home/-/testray/case_results/2089657465)
Semantic module returned a failure in `/apps/client-extension/client-extension-type-api`  file 
```
  [exec] com.liferay.semantic.versioning.SemanticVersioningTest > testSemanticVersioning[/apps/client-extension/client-extension-type-api] FAILED
     [exec]     java.lang.AssertionError:   PACKAGE_NAME                                       DELTA      CUR_VER    BASE_VER   REC_VER    WARNINGS  
     [exec]     = ================================================== ========== ========== ========== ========== ==========
     [exec]     * com.liferay.client.extension.type                  MINOR      3.0.2      3.0.2      3.1.0      VERSION INCREASE REQUIRED
     [exec]     	+   interface  com.liferay.client.extension.type.ThemeSpritemapCET
     [exec]     		+   access     abstract
     [exec]     		+   implements com.liferay.client.extension.type.CET
     [exec]     		+   method     getBaseURL()
     [exec]     			+   access     abstract
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getCompanyId()
     [exec]     			+   access     abstract
     [exec]     			+   return     long
     [exec]     		+   method     getDescription()
     [exec]     			+   access     abstract
     [exec]     			+   annotated  com.liferay.client.extension.type.annotation.CETProperty
     [exec]     				+   property   name='description'
     [exec]     				+   property   type='string'
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getEditJSP()
     [exec]     			+   access     abstract
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getExternalReferenceCode()
     [exec]     			+   access     abstract
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getName(java.util.Locale)
     [exec]     			+   access     abstract
     [exec]     			+   annotated  com.liferay.client.extension.type.annotation.CETProperty
     [exec]     				+   property   name='name'
     [exec]     				+   property   type='string'
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getProperties()
     [exec]     			+   access     abstract
     [exec]     			+   return     java.util.Properties
     [exec]     		+   method     getSourceCodeURL()
     [exec]     			+   access     abstract
     [exec]     			+   annotated  com.liferay.client.extension.type.annotation.CETProperty
     [exec]     				+   property   defaultValue='https://www.liferay.com'
     [exec]     				+   property   name='sourceCodeURL'
     [exec]     				+   property   type='string'
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getStatus()
     [exec]     			+   access     abstract
     [exec]     			+   return     int
     [exec]     		+   method     getType()
     [exec]     			+   access     abstract
     [exec]     			+   annotated  com.liferay.client.extension.type.annotation.CETProperty
     [exec]     				+   property   name='type'
     [exec]     				+   property   type='string'
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getTypeSettings()
     [exec]     			+   access     abstract
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     getURL()
     [exec]     			+   access     abstract
     [exec]     			+   annotated  com.liferay.client.extension.type.annotation.CETProperty
     [exec]     				+   property   defaultValue=''
     [exec]     				+   property   name='url'
     [exec]     				+   property   type='string'
     [exec]     				+   property   url=true
     [exec]     			+   return     java.lang.String
     [exec]     		+   method     hasProperties()
     [exec]     			+   access     abstract
     [exec]     			+   return     boolean
     [exec]     		+   method     isReadOnly()
     [exec]     			+   access     abstract
     [exec]     			+   return     boolean
     [exec]     		+   annotated  com.liferay.client.extension.type.annotation.CETType
     [exec]     			+   property   description='This is a description.'
     [exec]     			+   property   name='themeSpritemap'
     [exec]     		+   annotated  org.osgi.annotation.versioning.ProviderType
     [exec]     [Baseline Warning] Bundle Version Change Recommended: 5.1.0
     [exec]         at org.junit.Assert.fail(Assert.java:88)
```

2.  com.liferay.frontend.taglib.servlet.taglib.ComponentTagTest
```
testDoEndTag: org.junit.ComparisonFailure: expected:<...space","spritemap":"[\/clay\/icons.svg]"}), { destroyOnNavi...> but was:<...space","spritemap":"[]"}), { destroyOnNavi...>
```

File: modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/admin/edit_theme_spritemap.jsp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+<%@ include file="/admin/init.jsp" %>
+
+<%
+EditClientExtensionEntryPartDisplayContext<ThemeSpritemapCET> editClientExtensionEntryPartDisplayContext = (EditClientExtensionEntryPartDisplayContext)renderRequest.getAttribute(ClientExtensionAdminWebKeys.EDIT_CLIENT_EXTENSION_ENTRY_PART_DISPLAY_CONTEXT);
+
+ThemeSpritemapCET themeSpritemapCET = editClientExtensionEntryPartDisplayContext.getCET();
+%>
+
+<aui:input label="url" name="url" type="text" value="<%= themeSpritemapCET.getURL() %>" />
```
Comment (dsanz) [inline modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/admin/edit_theme_spritemap.jsp]: `required=true`?

File: modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/display/context/util/CETLabelUtil.java [hunk 2]
```diff
@@ -50,6 +50,11 @@ else if (Objects.equals(
 
 			return "css";
 		}
+		else if (Objects.equals(
+					type, ClientExtensionEntryConstants.TYPE_THEME_SPRITEMAP)) {
+
+			return "svg";
```
Comment (dsanz) [inline modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/display/context/util/CETLabelUtil.java:56]: Small detail: as we're overriding the theme SVG spritemap, this label may say something in line with the other `theme-*` types of extension (lines 71 and 76), kinda `theme-svg` which would be translated as one of "Theme SVG Spritemap", "Theme SVG" or "Theme spritemap" in the language.properties

File: modules/apps/layout/layout-admin-web/src/main/resources/META-INF/resources/layout_set/look_and_feel.jsp [hunk 3]
```diff
@@ -37,6 +37,26 @@ LayoutSet selLayoutSet = layoutsAdminDisplayContext.getSelLayoutSet();
...
+	/>
+
+	<p>
+		<liferay-ui:message key="use-this-client-extension-to-fully-replace-the-default-spritemap-contained-in-the-theme" />
+	</p>
+
+	<div>
+		<react:component
+			module="js/layout/look_and_feel/ThemeSpritemapCETsConfiguration"
+			props="<%= layoutLookAndFeelDisplayContext.getThemeSpritemapCETConfigurationProps(Layout.class.getName(), selLayoutSet.getLayoutSetId()) %>"
```
Comment (dsanz) [inline modules/apps/layout/layout-admin-web/src/main/resources/META-INF/resources/layout_set/look_and_feel.jsp:55]: This shall be `LayoutSet.class.getName()`if we follow the pattern in line 65, right?

**Comment by dsanz** [comment]
LGTM,great job @bryceosterhaus  pls see some minor comments. 

About the failures, we'll also need to run `gw baseline` on the type-api module and find out a solution for [this template](https://github.com/liferay/liferay-portal/blob/3df4b9eb2b7911fcdc331b046e38571e80e0346e/modules/apps/frontend-taglib/frontend-taglib/src/test/resources/com/liferay/frontend/taglib/servlet/taglib/dependencies/component_tag.html#L7)

**Comment by bryceosterhaus** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
I'm taking a look 🕵🏽

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
I'm taking a look again 🕵🏽

**Comment by mateusralv** [comment]
Hi @bryceosterhaus I analyzed the new results (https://github.com/liferay-frontend/liferay-portal/pull/2887?notification_referrer_id=NT_kwDOAqhMcrM0OTYyMzQ4NjMwOjQ0NTg0MDUw#issuecomment-1340992500) and one failure seems related to your PR.

[comparison run](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1670422731442&testrayRunIdA=2106004744&testrayRunIdB=2108003361)

### Unique failure 

[modules-semantic-versioning-jdk8/0/0](https://testray.liferay.com/home/-/testray/case_results/2108020497)
```
 [exec] com.liferay.semantic.versioning.SemanticVersioningTest > testSemanticVersioning[/apps/client-extension/client-extension-api] FAILED
     [exec]     java.lang.AssertionError:   PACKAGE_NAME                                       DELTA      CUR_VER    BASE_VER   REC_VER    WARNINGS  
     [exec]     = ================================================== ========== ========== ========== ========== ==========
     [exec]     * com.liferay.client.extension.constants             MINOR      2.0.0      2.0.0      2.1.0      VERSION INCREASE REQUIRED
     [exec]     	<   class      com.liferay.client.extension.constants.ClientExtensionEntryConstants
     [exec]     		+   field      TYPE_THEME_SPRITEMAP
     [exec]     			+   access     final
     [exec]     			+   access     static
     [exec]     			+   constant   themeSpritemap
     [exec]     			+   return     java.lang.String
     [exec]     [Baseline Warning] Bundle Version Change Recommended: 7.1.0
     [exec]         at org.junit.Assert.fail(Assert.java:88)
     [exec]         at org.junit.Assert.assertTrue(Assert.java:41)
     [exec]         at org.junit.Assert.assertFalse(Assert.java:64)
```


### Not related

1. modules-compile-jdk8/0/2
2. portal-license-smoke-jdk8/0/0
3. LocalFile.SystemObjectAccount#CanAssociateAnyTypeOfAccount
4. com.liferay.portal.security.membership.policy.role.test.RoleMembershipPolicyRolesTest
5. com.liferay.portal.security.membership.policy.usergroup.test.UserGroupMembershipPolicyMembershipsTest
6. LocalFile.MakeRelationshipWithSystemObject#CanDeleteCustomObjectEntryRelaatedToSystemObjectByManyToMany

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
lets try baseline one more time...

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by john-co** [comment]
Just started reviewing :)

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated. :heavy_check_mark: 

Ok to manually forward from QA perspective. Pending dev approval.

**Comment by john-co** [comment]
[Testray Comparison](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=2108691035&testrayRunIdA=2111268819&statusB=2)

Mostly all "unique" failing tests are either recently failing on other PRs or sporadically failing based on history.

**Comment by bryceosterhaus** [comment]
Manual forward at https://github.com/brianchandotcom/liferay-portal/pull/127532

**Pull request closed without merging**